### PR TITLE
refactor(components): De-hardcoded local output paths.

### PR DESCRIPTION
**Pull request by Ark-kun**
Preparation for the future storage system.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/580)

### CODE REVIEW

File: components/dataflow/predict/src/predict.py [hunk 1]
```diff
@@ -60,6 +61,14 @@ def parse_arguments():
                       type=int,
                       default=32,
                       help='Batch size used in prediction.')
+  parser.add_argument('--prediction-results-uri-pattern-output-path',
```
Comment (hongye-sun) [inline components/dataflow/predict/src/predict.py]: I probably missed some context here. Why is this needed? Does it add burden to component author to add each output as a command line argument?

**Comment by gaoning777** [comment]
Are you planning to close this issue? since you did not try to resolve hongye's concern, should we close it?

File: components/dataflow/predict/src/predict.py (hunk 1 above)
Comment (Ark-kun) [reply components/dataflow/predict/src/predict.py]: The industry is moving away from Docker towards open and standardized container support and so do our users and customers.
We have received bug reports from users that our pipelines fail on non-Docker Kubernetes clusters.
While making Argo work on a non-Docker cluster is trivial, it's much harder with Pipelines.
One particular problem is with local artifact paths. There are specific system-imposed requirements for where the artifacts can and cannot be stored. More so, this depends on the Argo executor being used in the cluster.
De-hardcoding the local artifact paths seems to be the only option for keeping the pipelines portable.
Usually the programs are already written without any paths hard-coded.
If the program did not follow the best practices, the author only need to add one line to the program to resolve the problem. This seem to be a small price where the alternative is a component that only work on some clusters.

**Comment by Ark-kun** [comment]
We're starting to get more users report problems caused by incompatible artifact output paths: https://github.com/kubeflow/pipelines/issues/1654

The industry is moving away from Docker towards open and standardized container support and so do our users and customers.
We have received bug reports from users that our pipelines fail on non-Docker Kubernetes clusters.
While making Argo work on a non-Docker cluster is trivial, it's much harder with Pipelines.
One particular problem is with local artifact paths. There are specific system-imposed requirements for where the artifacts can and cannot be stored. More so, this depends on the Argo executor being used in the cluster.
De-hardcoding the local artifact paths seems to be the only option for keeping the pipelines portable.
Usually the programs are already written without any paths hard-coded.
If the program did not follow the best practices, the author only need to add one line to the program to resolve the problem. This seem to be a small price where the alternative is a component that only work on some clusters.

**Comment by xinbinhuang** [comment]
@Ark-kun One quick question: does it mean that component should not utilize `local` output path but external storage system for "artifacts" (i.e. not the argo artifacts) that user want to persist?

**Comment by Ark-kun** [comment]
> @Ark-kun One quick question: does it mean that moving forward, components should not utilize `local` output path but external storage system for "artifacts" (i.e. not the argo artifacts) that user want to persist?

No. The components should still output data by writing it to local paths unless they output 100GB of data.

But components should not hardcode those output paths in the code. The local paths should be given by the system.

Does this make sense?

**Comment by Bobgy** [comment]
/retest

Looks like we need to fix this flakiness

**Comment by Bobgy** [comment]
/lgtm
/approve
Thanks @Ark-kun
